src/keyring/KeyringController.test.ts

```diff
@@ -1094,4 +1096,302 @@ describe('KeyringController', () => {
+          value: '1337',
+        },
+      ];
+      const account = await ledgerKeyring.getDefaultAccount();
+
+      const signature = keyringController.signTypedMessage(
+        { data: typedMsgParams, from: account },
+        SignTypedDataVersion.V1,
+      );
+
+      await expect(signature).rejects.toThrow(
+        'Ledger: Only version 4 of typed data signing is supported',
+      );
+
+      const signature2 = keyringController.signTypedMessage(
+        { data: typedMsgParams, from: account },
+        SignTypedDataVersion.V3,
+      );
+
+      await expect(signature2).rejects.toThrow(
+        'Ledger: Only version 4 of typed data signing is supported',
+      );
+    });
+
+    it('should sign typed data for V4 with Ledger Keyring', async () => {
+      const msgParams = {
+        domain: {
+          chainId: 1,
+          name: 'Ether Mail',
+          verifyingContract: '0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC',
+          version: '1',
+        },
+        message: {
+          contents: 'Hello, Bob!',
+          attachedMoneyInEth: 4.2,
+          from: {
+            name: 'Cow',
+            wallets: [
+              '0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826',
+              '0xDeaDbeefdEAdbeefdEadbEEFdeadbeEFdEaDbeeF',
+            ],
+          },
+          to: [
+            {
+              name: 'Bob',
+              wallets: [
+                '0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB',
+                '0xB0BdaBea57B0BDABeA57b0bdABEA57b0BDabEa57',
+                '0xB0B0b0b0b0b0B000000000000000000000000000',
+              ],
+            },
+          ],
+        },
+        primaryType: 'Mail',
+        types: {
+          EIP712Domain: [
+            { name: 'name', type: 'string' },
+            { name: 'version', type: 'string' },
+            { name: 'chainId', type: 'uint256' },
+            { name: 'verifyingContract', type: 'address' },
+          ],
+          Group: [
+            { name: 'name', type: 'string' },
+            { name: 'members', type: 'Person[]' },
+          ],
+          Mail: [
+            { name: 'from', type: 'Person' },
+            { name: 'to', type: 'Person[]' },
+            { name: 'contents', type: 'string' },
+          ],
+          Person: [
+            { name: 'name', type: 'string' },
+            { name: 'wallets', type: 'address[]' },
+          ],
+        },
+      };
+
+      const confirmSignatureStub = sinon.stub(ledgerKeyring, 'signTypedData');
+
+      confirmSignatureStub.resolves(
+        '0xafb6e247b1c490e284053c87ab5f6b59e219d51f743f7a4d83e400782bc7e4b9479a268e0e0acd4de3f1e28e4fac2a6b32a4195e8dfa9d19147abe8807aa6f6400',
+      );
+
+      const account = await ledgerKeyring.getDefaultAccount();
+      const signature = await keyringController.signTypedMessage(
+        { data: JSON.stringify(msgParams), from: account },
+        SignTypedDataVersion.V4,
+      );
+      const recovered = recoverTypedSignature_v4({
+        data: msgParams as any,
```

Comment:
Why are we typecasting here? Is it because `eth-sig-util` doesn't have proper types (or we're using a version that doesn't have proper types)?

Reply:
@mcmire I have removed the `any` casting and added the proper data types in [ebc394a](https://github.com/MetaMask/controllers/pull/862/commits/ebc394acad183aec4518a5b2bc2ffe825b4c2088)

Reply:
Oops, forgot to follow up, this looks good to me.